Extract students-in-filter component from overview and add to STAR reading / math pages

jonchar opened the pull request:
Opening this to get feedback while I work out what's going on in #375. What's happening is that the number of students being filtered via the `SlicePanel` doesn't match up with the number reported in the visualization.

I couldn't reproduce a large discrepancy when running locally, but added a div to display information about the number of students the `SlicePanel` is selecting. I'm not sure whether it should go in `slice_panels.js` or `star_charts_page.js` depending on if the indicator is relevant in other views. Suggestions welcome how to fit it in style-wise of course :smile: . I assume this number should be the same as what gets displayed in the visualization generated by `StarChartsPage`.

This shows the indicator below the columns with 64 students. Just below the cutoff it reads there are 56 students in the visualization.

![slicepanel](https://cloud.githubusercontent.com/assets/10509931/15619856/dd3f53fa-2424-11e6-8b50-fa49fcc6ff30.png)

kevinrobinson commented:
@jonchar Thanks!  This might be helpful for you too: https://github.com/studentinsights/studentinsights/issues/375#issuecomment-222262168

jonchar commented:
I just finished extracting the students-in-filter parts from the `school_overview_page.js` (see https://github.com/studentinsights/studentinsights/blob/master/app/assets/javascripts/school_overview/school_overview_page.js#L212) into its own component. I named the new component `slice_buttons.js` (any suggestions for a clearer name are welcome). I also removed the count that I had previously added into `slice_panels.js` for this PR (see above).

I also added the newly extracted component to `star_charts_page.js` so we now have a students-in-filter count for the STAR Reading / Math pages. This should address #375 aside from further documenting the filtering edge cases @ianaroot found. Not sure if that warrants a separate issue.

Someone should probably have a look at this before merging just to be sure my React is on par :smile: .

jonchar commented:
Realized I had missed the code for using the ESC key to clear the filters. Just moved that code to `slice_buttons.js` as well. Thanks @erose !

erose commented:
Jon is this still a WIP or would you like it to get reviewed? I took away the `needs-review` label after we talked about the ESC key stuff, but reapply as needed.

jonchar commented:
I'm believe I properly moved the ESC key functionality into `slice_buttons.js` and out of `school_overview_page.js ` & `star_charts_page.js`. The buttons work as expected on both the overview and STAR pages. I'm re-applying the `needs-review` however because someone else should have a look since it's my first attempt at making a new component. Other than that I think it's good to go.

kevinrobinson commented on app/assets/javascripts/components/slice_buttons.js:
```diff
@@ -0,0 +1,112 @@
+      students: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
+      filters: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
+      filtersHash: React.PropTypes.func.isRequired,
+      activeFiltersIdentifier: React.PropTypes.func.isRequired,
```
I think `SliceButtons` only needs the values for `filtersHash` and `activeFiltersIdentifier`.  So you can have the call sites evaluate those functions and then pass the plain value into `SliceButtons` (the same way that `students` works).

Like this:
```
createEl(SliceButtons, {
  students: this.filteredStudents(),
  filters: this.state.filters,
  filtersHash: this.filtersHash(),
  activeFiltersIdentifier: this.activeFiltersIdentifier(),
  onClearFilters: this.onClearFilters
})
```

kevinrobinson commented on app/assets/javascripts/components/star_charts_page.js:
```diff
@@ -76,6 +85,14 @@ $(function() {
+
+    //onKeyDown: function(e) {
+    //  if (e.keyCode == 27) this.clearFilters();
+    //},
```
Leftover comments here.

kevinrobinson commented on app/assets/javascripts/components/star_charts_page.js:
```diff
@@ -186,7 +210,7 @@ $(function() {
         dom.div({},
           dom.span({}, 'Students: ', props.students.length),
-          dom.span({ style: { paddingLeft: 10 }}, 'Data points: ', (props.assessments || []).length)));
+          dom.span({ style: { paddingLeft: 10 }}, '# Data points: ', (props.assessments || []).length)));
```
Using the symbol might be cryptic, how about "Number of data points" if that fits?

kevinrobinson commented on app/assets/javascripts/school_overview/school_overview_page.js:
```diff
@@ -28,13 +29,15 @@
+    // Moved to slice_buttons.js
+    //componentWillUnmount: function() {
+    //  $(document).off('keydown', this.onKeyDown);
+    //},
```
We can remove these comments now, right?

kevinrobinson commented on Gemfile.lock:
```diff
@@ -328,5 +328,8 @@ DEPENDENCIES
+
 BUNDLED WITH
-   1.11.2
+   1.12.4
```
Is this intentional?  Unless it's necessary let's split out upgrades to dependencies into a separate PR.

kevinrobinson commented:
@jonchar Awesome!  Could you also drop in screenshots of what this looks like?  The comments are mostly about cleanup so should be straightforward to finish off this great improvement!

jonchar commented:
Thanks for the comments @kevinrobinson ! I think I've addressed them all. Here is a screenshot of how the `SliceButtons` look on the STAR Reading page.

![slicebuttons_on_star_page](https://cloud.githubusercontent.com/assets/10509931/16313996/dfe6ce6e-3949-11e6-8be6-b8721d2e6bf0.png)

kevinrobinson commented:
@jonchar awesome! 😄
